File: oop1.py
from abc import ABC, abstractmethod

class BankAccount(ABC):
    def __init__(self, account_number, initial_balance):
        self.__account_number = account_number  # Encapsulation (private) for protected only _
        self.__balance = initial_balance

    @abstractmethod
    def deposit(self, amount):
        pass

    @abstractmethod
    def withdraw(self, amount):
        pass

    def get_account_info(self):
        return f"Account Number: {self.__account_number}, Balance: {self.__balance}"

    def _update_balance(self, amount):  # Protected method
        self.__balance += amount

    def get_balance(self):
        return self.__balance

# Using Inheritance
class SavingAccount(BankAccount):
    def __init__(self, account_number, initial_balance, interest_rate):
        super().__init__(account_number, initial_balance)
        self.__interest_rate = interest_rate

    def deposit(self, amount):
        self._update_balance(amount)  # Encapsulated method

    def withdraw(self, amount):
        if amount <= self.get_balance():
            self._update_balance(-amount)  # Withdraw within balance limit
        else:
            print("Insufficient Funds.")

    def calculate_interest(self):
        return self.get_balance() + (self.get_balance() * self.__interest_rate / 100)  # Corrected Calculation

class CheckingAccount(BankAccount):
    def __init__(self, account_number, initial_balance, overdraft_limit):
        super().__init__(account_number, initial_balance)
        self.__overdraft_limit = overdraft_limit

    def deposit(self, amount):
        self._update_balance(amount)

    def withdraw(self, amount):
        if amount <= self.get_balance() + self.__overdraft_limit:
            self._update_balance(-amount)
        else:
            print("Overdraft limit exceeded.")

# Testing the classes
savings = SavingAccount("123456", 5000, 5)
savings.deposit(1000)
print(savings.get_account_info())
print(f"Interest Earned: {savings.calculate_interest()}")

checking = CheckingAccount("654321", 3000, 1000)
checking.withdraw(3500)
print(checking.get_account_info())

File: oop1.java
// Abstract Base Class
abstract class BankAccount {
    private String accountNumber;  // Encapsulation (private)
    protected double balance;  // Protected for controlled access in subclasses

    public BankAccount(String accountNumber, double initialBalance) {
        this.accountNumber = accountNumber;
        this.balance = initialBalance;
    }

    public abstract void deposit(double amount);

    public abstract void withdraw(double amount);

    public String getAccountInfo() {
        return "Account Number: " + accountNumber + ", Balance: " + balance;
    }

    protected void updateBalance(double amount) { // Protected method
        this.balance += amount;
    }

    public double getBalance() {
        return balance;
    }
}

// Using Inheritance
class SavingAccount extends BankAccount {
    private double interestRate; // Encapsulation (private)

    public SavingAccount(String accountNumber, double initialBalance, double interestRate) {
        super(accountNumber, initialBalance);
        this.interestRate = interestRate;
    }

    @Override
    public void deposit(double amount) {
        updateBalance(amount); // Encapsulated method
    }

    @Override
    public void withdraw(double amount) {
        if (amount <= getBalance()) {
            updateBalance(-amount);
        } else {
            System.out.println("Insufficient Funds.");
        }
    }

    public double calculateInterest() {
        return getBalance() + (getBalance() * interestRate / 100); // Corrected Calculation
    }
}

// Another Child Class (Polymorphism)
class CheckingAccount extends BankAccount {
    private double overdraftLimit;

    public CheckingAccount(String accountNumber, double initialBalance, double overdraftLimit) {
        super(accountNumber, initialBalance);
        this.overdraftLimit = overdraftLimit;
    }

    @Override
    public void deposit(double amount) {
        updateBalance(amount);
    }
    
    
    @Override
    public void withdraw(double amount) {
        if (amount <= getBalance() + overdraftLimit) {
            updateBalance(-amount);
        } else {
            System.out.println("Overdraft limit exceeded.");
        }
    }
}

// Demonstrating functionality
public class BankSystem {
    public static void main(String[] args) {
        SavingAccount savings = new SavingAccount("123456", 5000, 5);
        savings.deposit(1000);
        System.out.println(savings.getAccountInfo());
        System.out.println("Interest Earned: " + savings.calculateInterest());

        CheckingAccount checking = new CheckingAccount("654321", 3000, 1000);
        checking.withdraw(3500);
        System.out.println(checking.getAccountInfo());
    }
}